app/src/main/java/io/gnosis/safe/ui/transaction/TransactionViewHolderFactory.kt

```diff
@@ -0,0 +1,103 @@
+    CHANGE_MASTERCOPY, CHANGE_MASTERCOPY_QUEUED, SETTINGS_CHANGE, SETTINGS_CHANGE_QUEUED, TRANSFER, TRANSFER_QUEUED
+}
+
+class TransactionViewHolderFactory : BaseFactory<BaseTransactionViewHolder<Transaction>, Transaction>() {
+
+    @Suppress("UNCHECKED_CAST")
+    override fun newViewHolder(viewBinding: ViewBinding, viewType: Int): BaseTransactionViewHolder<Transaction> =
+        when (viewType) {
+            TransactionViewType.CHANGE_MASTERCOPY.ordinal -> ChangeMastercopyViewHolder(viewBinding as ItemTxChangeMastercopyBinding)
+            TransactionViewType.CHANGE_MASTERCOPY_QUEUED.ordinal -> ChangeMastercopyQueuedViewHolder(viewBinding as ItemTxQueuedChangeMastercopyBinding)
+            TransactionViewType.SETTINGS_CHANGE.ordinal -> SettingsChangeViewHolder(viewBinding as ItemTxSettingsChangeBinding)
+            TransactionViewType.SETTINGS_CHANGE_QUEUED.ordinal -> SettingsChangeQueuedViewHolder(viewBinding as ItemTxQueuedSettingsChangeBinding)
+            TransactionViewType.TRANSFER.ordinal -> TransferViewHolder(viewBinding as ItemTxTransferBinding)
+            TransactionViewType.TRANSFER_QUEUED.ordinal -> TransferQueuedViewHolder(viewBinding as ItemTxQueuedTransferBinding)
+            else -> throw UnsupportedViewType(javaClass.name)
+        } as BaseTransactionViewHolder<Transaction>
+
+
+    override fun layout(layoutInflater: LayoutInflater, parent: ViewGroup, viewType: Int): ViewBinding =
+        when (viewType) {
+            TransactionViewType.CHANGE_MASTERCOPY.ordinal -> ItemTxChangeMastercopyBinding.inflate(layoutInflater, parent, false)
+            TransactionViewType.CHANGE_MASTERCOPY_QUEUED.ordinal -> ItemTxQueuedChangeMastercopyBinding.inflate(layoutInflater, parent, false)
+            TransactionViewType.SETTINGS_CHANGE.ordinal -> ItemTxSettingsChangeBinding.inflate(layoutInflater, parent, false)
+            TransactionViewType.SETTINGS_CHANGE_QUEUED.ordinal -> ItemTxQueuedSettingsChangeBinding.inflate(layoutInflater, parent, false)
+            TransactionViewType.TRANSFER.ordinal -> ItemTxTransferBinding.inflate(layoutInflater, parent, false)
+            TransactionViewType.TRANSFER_QUEUED.ordinal -> ItemTxQueuedTransferBinding.inflate(layoutInflater, parent, false)
+            else -> throw UnsupportedViewType(javaClass.name)
+        }
+
+    override fun viewTypeFor(item: Transaction): Int =
+        when (item) {
+            is Transaction.ChangeMastercopy -> TransactionViewType.CHANGE_MASTERCOPY
+            is Transaction.SettingsChange -> TransactionViewType.SETTINGS_CHANGE
+            is Transaction.Transfer -> TransactionViewType.TRANSFER
+            is Transaction.ChangeMastercopyQueued -> TransactionViewType.CHANGE_MASTERCOPY_QUEUED
+            is Transaction.SettingsChangeQueued -> TransactionViewType.SETTINGS_CHANGE_QUEUED
+            is Transaction.TransferQueued -> TransactionViewType.TRANSFER_QUEUED
+            else -> throw UnsupportedViewType(javaClass.name)
+        }.ordinal
```

Comment:
Do we risk breaking the code when (accidentally) reordering the enum? Should we detect this with a unit test?

Reply:
There is no risk of that. `ordinal` returns the index within the enum whatever it may be at the time of compilation. As long as we use `ordinal` instead of hard coded indexes everything should be safe. 
Unit tests would be a bit hard at the moment. `layout` calls static methods from `ViewBinding` classes. 
I suppose I could try testing, see that the instances of viewHolders returned conform to the expected type, etc.

Reply:
@biafra23 I think that the tests might not work. `ViewBinding` doesn't seem to fly with unit tests. I am afraid this might not work on build kite. I suppose when we start with UI testing this would be feasible to test.